<launch>
    <!-- Source name
    alsa_input.usb-SEEED_ReSpeaker_4_Mic_Array__UAC1.0_-00.multichannel-input
    pulse_audio_source" default="alsa_input.usb-IntRoLab_16SoundsUSB_Audio_2.0-00.multichannel-input
    alsa_output.pci-0000_00_1f.3.analog-stereo

    <rosparam param="channel_map">
            [front-left, front-right, rear-left, rear-right, front-center, lfe, side-left, side-right, aux0, aux1, aux2, aux3, aux4, aux5, aux6, aux7]
    </rosparam>
    -->

    <!-- ARGUMENTS -->
    <arg name="audio_capture" default="false"/>
    <arg name="merge_speech_noise" default="false"/>
    <arg name="calibration_active" default="false"/>
    <arg name="egonoise_active" default="false"/>
    <arg name="audio_listen" default="false"/>

    <arg name="pulse_audio_sink" default="alsa_output.pci-0000_00_1f.3.analog-stereo"/>
    <arg name="pulse_audio_source" default="alsa_input.usb-SEEED_ReSpeaker_4_Mic_Array__UAC1.0_-00.multichannel-input"/>
    <arg name="use_echo_cancellation" default="false"/>
    <arg name="echo_cancellation_dest" default="false"/>
    <arg name="dict_path" default="/home/pierre-olivier/catkin_ws/src/egonoise/src/database/"/>
    <arg name="channel_count_in" default="6"/>
    <arg name="channel_keep" default="[1,2,3,4]"/>
    <arg name="channel_count_out" default="1"/>
    <arg name="sampling_frequency" default="16000"/>
    <arg name="frame_sample_count" default="16000"/>
    <arg name="frame_size" default="512"/>
    <arg name="format" default="signed_32"/>


    <!-- AUDIO CAPTURE -->
    <node pkg="audio_utils" type="capture_node" name="capture_node" output="screen" launch-prefix="bash -c 'sleep 5; $0 $1 $2 $3 $4 $5 $6' " if="$(arg audio_capture)">
        <param name="backend" value="pulse_audio"/>
        <param unless="$(arg use_echo_cancellation)" name="device" value="$(arg pulse_audio_source)"/>
        <param if="$(arg use_echo_cancellation)" name="device" value="$(arg echo_cancellation_dest)_ec"/>
        <param name="format" value="$(arg format)"/>
        <param name="channel_count" value="$(arg channel_count_in)"/>
        <param name="sampling_frequency" value="$(arg sampling_frequency)"/>
        <param name="frame_sample_count" value="$(arg frame_sample_count)"/>
        <param name="merge" value="false"/>
        <param name="gain" value="1.0"/>
    </node>

    <!-- CALIBRATION -->
    <node pkg="egonoise" type="calibration_node.py" name="calibration_node" output="screen" if="$(arg calibration_active)">
        <param name="input_format" value="$(arg format)"/>
        <param name="sampling_frequency" value="$(arg sampling_frequency)"/>
        <param name="database_path" value="$(arg dict_path)"/>

        <rosparam param="channel_keep">[1, 2, 3, 4]</rosparam>
    </node>

    <!-- EGONOISE -->
    <node pkg="egonoise" type="egonoise_node.py" name="egonoise_node" output="screen" if ="$(arg egonoise_active)">
        <param name="input_format" value="$(arg format)"/>
        <param name="output_format" value="$(arg format)"/>
        <param name="dict_path" value="$(arg dict_path)"/>
        <param name="frame_size" value="$(arg frame_size)"/>
        <param name="channel_count" value="$(arg channel_count_in)"/>

        <rosparam param="channel_keep">[1, 2, 3, 4]</rosparam>
    </node>

    <!-- AUDIO LISTEN -->
    <node pkg="audio_utils" type="playback_node" name="playback_node" respawn="true" if="$(arg audio_listen)">
        <param name="backend" value="pulse_audio"/>
        <param name="device" value="$(arg pulse_audio_sink)"/>
        <param name="format" value="$(arg format)"/>
        <param name="channel_count" value="$(arg channel_count_out)"/>
        <param name="sampling_frequency" value="$(arg sampling_frequency)"/>
        <param name="frame_sample_count" value="$(arg frame_sample_count)"/>

    </node>


</launch>